### PR TITLE
Test whether projections can be configured

### PR DESCRIPTION
**Pull request by UFOMelkor**
Verifies #42

### CODE REVIEW

**Comment by codeliner** [comment]
Ok, so the failing test is a good thing?
Do you want to merge the test and then fix the bug in another PR?

**Comment by UFOMelkor** [comment]
Will try to fix tomorrow ;-)

**Comment by UFOMelkor** [comment]
Finally got some time to have a deeper look into this issue.
Did some refactoring, but this will need at least some cleaning up.

**Comment by UFOMelkor** [comment]
I removed all aliases for projections and read models.
Access to projections, read models and projection managers is provided only thought the dedicated service locators.

@Miljar @kejwmen @gquemener @lunetics Would be great if one of you (or any other people which use this bundle) have time to review this carefully.

File: src/DependencyInjection/Compiler/ProjectorPass.php [hunk 1]
```diff
@@ -18,67 +18,123 @@
...
 use Symfony\Component\DependencyInjection\Reference;
-use Symfony\Component\DependencyInjection\ServiceLocator;
 
 final class ProjectorPass implements CompilerPassInterface
 {
-    public function process(ContainerBuilder $container)
+    public function process(ContainerBuilder $container): void
     {
-        $projectors = $container->findTaggedServiceIds(ProophEventStoreExtension::TAG_PROJECTION);
+        $projectorsIds = array_keys($container->findTaggedServiceIds(ProophEventStoreExtension::TAG_PROJECTION));
```
[the remark this answers is not in the document]
Comment (gquemener) [reply src/DependencyInjection/Compiler/ProjectorPass.php]: This class could be named `ProjectionPass` as well (or even better something like `RegisterProjectionsPass`.

File: src/DependencyInjection/Compiler/ProjectorPass.php [hunk 2]
```diff
@@ -18,67 +18,123 @@
...
+
+        if (! $container->hasDefinition('prooph_event_store.projection_read_models_locator')
```
Comment (gquemener) [inline src/DependencyInjection/Compiler/ProjectorPass.php]: These service ids could be stored in static properties.

File: src/DependencyInjection/Compiler/ProjectorPass.php [hunk 3]
```diff
@@ -18,67 +18,123 @@
...
+     * @param string $serviceId The id of the service whose tag is verified
+     * @param array $tag The actual tag
+     * @param string $attributeName The attribute that has to be available in the tag
+     * @throws RuntimeException if the attribute is not available in the tag
+     */
+    private static function assertProjectionTagHasAttribute(string $serviceId, array $tag, string $attributeName): void
+    {
+        if (! isset($tag[$attributeName])) {
+            throw new RuntimeException(sprintf(
+                '"%s" attribute is missing from on "%s" tagged service "%s"',
```
Comment (gquemener) [inline src/DependencyInjection/Compiler/ProjectorPass.php]: There's a typo here.

Correct string should be stg like `"%s" attribute is missing from tag "%s" on service "%s".`

**Comment by UFOMelkor** [comment]
Thank you @gquemener 

> You are trying to provide projection configuration through service tags, in addition of bundle configuration, right?

Yes

> On the other hand, the DI extension code is quite complex and suffers from some inconsistency which makes it harder to understand.

Yes, but as you said out of scope.
We can (and should) provide another PR to smooth the bundle without changing any features.

> Some code is dead (see lines 144 and 145 for example).

Can you name the file?

**Comment by gquemener** [comment]
Sure, the dead code is here: https://github.com/prooph/event-store-symfony-bundle/blob/master/src/DependencyInjection/ProophEventStoreExtension.php?utf8=%E2%9C%93#L164-L165

**Comment by UFOMelkor** [comment]
> Sure, the dead code is here: https://github.com/prooph/event-store-symfony-bundle/blob/master/src/DependencyInjection/ProophEventStoreExtension.php?utf8=%E2%9C%93#L164-L165
Good catch!

**Comment by gquemener** [comment]
Any news/blockers on this PR @UFOMelkor ?

**Comment by UFOMelkor** [comment]
I've got time today to test it with proophessor-do-symfony and made one PR because of an error I got: https://github.com/prooph/proophessor-do-symfony/pull/25
While writing the changelog for the next PR i've thought a little bit more about why this error occured: In this PR I changed the configuration of the projections. I always thought, that the service id of the projections has to be passed to the configuration. But when looking again at [the old code](https://github.com/prooph/event-store-symfony-bundle/blob/3816dceae2c0d223b58e07731b157a346e8cd642/src/DependencyInjection/ProophEventStoreExtension.php#L134) a FQCN has to be passed to the configuration.

This would be a BC break, but 
 - would make the configuration more uniform
 - would be easy to fix in projects

Thoughts on this?

**Comment by gquemener** [comment]
Indeed, good catch.

Your idea would be to pass service id, instead of FQCN as `projection` value ?
That's a good idea to me.

In Symfony 3.4+ project, projection id will, often, be FQCN, so the change would be BC.

Unless a new major release is tagged, we need to find a solution to handle cases where `projection` IS a FQCN but NOT a service id.

Sounds like a lot of trouble for not much use cases (this bundle is not so much used out there, is it?).

*I would tag a 1.0.0 version*

**Comment by UFOMelkor** [comment]
@gquemener @enumag Can you have a look at https://github.com/prooph/event-store-symfony-bundle/pull/43/commits/ec1e4477f547920161367ecc6943bf3883e854ef ?

**Comment by UFOMelkor** [comment]
@enumag Which one?

File: src/DependencyInjection/Compiler/DeprecateFqcnProjectionsPass.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        $projections = array_map(function ($argument) {
+            if ($argument instanceof ServiceClosureArgument) {
+                $argument = $argument->getValues()[0];
+            }
+
+            return (string) $argument;
+        }, $container->getDefinition('prooph_event_store.projections_locator')->getArgument(0));
+        foreach ($projections as $eachProjectionId) {
```
Comment (enumag) [inline src/DependencyInjection/Compiler/DeprecateFqcnProjectionsPass.php]: newline before this foreach

File: test/DependencyInjection/Compiler/DeprecateFqcnProjectionsPassTest.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        $this->assertContainerBuilderHasService(BlackHoleProjection::class, BlackHoleProjection::class);
+    }
+
+    private function registerProjectionsServiceLocator(array $serviceIds): void
+    {
+        $serviceIds = array_map(function (string $serviceId) {
+            return new ServiceClosureArgument(new Reference($serviceId));
+        }, $serviceIds);
+        $this->container->setDefinition(
```
Comment (enumag) [inline test/DependencyInjection/Compiler/DeprecateFqcnProjectionsPassTest.php]: newline before this setDefinition

File: src/DependencyInjection/ProophEventStoreExtension.php [hunk 6]
```diff
@@ -39,104 +39,88 @@ public function getConfiguration(array $config, ContainerBuilder $container)
...
+        $definition->addTag('container.service_locator');
+        $container->setDefinition($id, $definition);
+    }
 
+    private static function collectProjectionsForLocators(
+        array $projections,
+        string $projectionManagerId,
+        array &$projectionManagerForProjectionsLocator,
+        array &$projectionsLocator,
+        array &$readModelsLocator
```
Comment (enumag) [inline src/DependencyInjection/ProophEventStoreExtension.php]: I'd get rid of the references in favor of this:

```
return [$projectionManagerForProjectionsLocator, $projectionsLocator, $readModelsLocator];

[$projectionManagerForProjectionsLocator, $projectionsLocator, $readModelsLocator] = self::collectProjectionsForLocators($projectionManagerConfig['projections'], $projectionManagerId,);
```

Comment (UFOMelkor) [reply src/DependencyInjection/ProophEventStoreExtension.php]: @gquemener your thoughts about out-parameters vs array-return?

**Comment by codeliner** [comment]
@gquemener 

> (this bundle is not so much used out there, is it?).

This bundle is definitely used in projects ;) But before v1.0 breaking changes are ok. Even better if communicated with deprecation errors :+1:

**Pull request merged**